File: fig_tauri/src-tauri/src/api/onboarding.rs
use fig_proto::fig::{
    OnboardingAction,
    OnboardingRequest,
};
use tokio::process::Command;

use super::{
    RequestResult,
    RequestResultImpl,
};

pub async fn onboarding(request: OnboardingRequest) -> RequestResult {
    match request.action() {
        OnboardingAction::InstallationScript => {
            // TODO(grant): Move install to a common lib and call directly
            match Command::new("fig")
                .args(&["_", "install", "--dotfiles", "--no-confirm"])
                .output()
                .await
            {
                Ok(_) => RequestResultImpl::success(),
                Err(err) => RequestResultImpl::error(err.to_string()),
            }
        },
        OnboardingAction::Uninstall => {
            // TODO(grant): Move uninstall to a common lib and call directly
            match Command::new("fig")
                .args(&["_", "uninstall", "--dotfiles"])
                .output()
                .await
            {
                Ok(_) => RequestResultImpl::success(),
                Err(err) => RequestResultImpl::error(err.to_string()),
            }
        },
        OnboardingAction::PromptForAccessibilityPermission
        | OnboardingAction::LaunchShellOnboarding
        | OnboardingAction::CloseAccessibilityPromptWindow
        | OnboardingAction::RequestRestart
        | OnboardingAction::CloseInputMethodPromptWindow => RequestResultImpl::error("Unimplemented"),
    }
}

File: typescript-api-bindings/src/onboarding.ts
import { onboardingActionFromJSON } from "./fig.pb";
import { sendOnboardingRequest } from "./requests";

export type OnboardingAction =
  | "INSTALLATION_SCRIPT"
  | "PROMPT_FOR_ACCESSIBILITY_PERMISSION"
  | "LAUNCH_SHELL_ONBOARDING"
  | "UNINSTALL"
  | "CLOSE_ACCESSIBILITY_PROMPT_WINDOW"
  | "REQUEST_RESTART"
  | "CLOSE_INPUT_METHOD_PROMPT_WINDOW";

export async function onboard(onboardingAction: OnboardingAction) {
  await sendOnboardingRequest({
    action: onboardingActionFromJSON(onboardingAction)
  });
}
